macOS sandboxed settings fixes

Be-ing opened the pull request:
Rebased successor of #3454

hile committed:
Fixes to moving settings folder to sandboxed location

This commit contains fixes for migrating old settings folder to
sandboxed path on macos when running as developer from git without
sandboxing enabled

Be-ing committed:
Sandbox: improve wording of dialog

Be-ing committed:
Sandbox: clarify comment

Be-ing committed:
Sandbox: do not bother users installing Mixxx >= 2.3.0 first

Holzhaus commented:
Is this feature completely? I can have a look at the code but tbh I have no idea about macOS.

Be-ing committed:
Sandbox: fix pre-commit issues

Be-ing commented:
This is ready for review. I will test both the signed and unsigned builds on macOS 10.13 but I cannot verify if it works on macOS 11.

Holzhaus commented:
Maybe we can ask on the forums?

Be-ing commented:
The check whether the process is sandboxed is not working.

Be-ing committed:
Sandbox: fix check whether sandboxing is enabled

Be-ing committed:
Sandbox: fix inverted if condition

Be-ing committed:
fix initialization of settings path

Be-ing committed:
Sandbox::migrateOldSettings reorganize code, improve comments

Be-ing committed:
Sandbox::migrateOldSettings reword dialog

Be-ing commented:
I have tested the following with both signed and unsigned builds and they all work:
1. Running Mixxx without any existing settings. New settings are created in the sandboxed path without prompting the user.
2. Running Mixxx with settings in the legacy path `~/Library/Application Support/Mixxx` and nothing in the sandboxed `~/Library/Containers/org.mixxx.mixxx`. The user is prompted for permission to access the old path and it is successfully moved into the sandboxed path.
3. Starting Mixxx again with settings in the sandboxed path. The settings in the sandbox are used.

I previously thought that the GH Actions builds for the `push` trigger on upstream branches had access to the secret variables for code signing but the builds for the `pull_request` trigger does not. I was wrong. The builds for both triggers have access to the secret variables for code signing. I had to push to my fork to get GH Actions to make unsigned builds.

Be-ing committed:
Sandbox: update #if conditions

esbrandt commented:
> I previously thought that the GH Actions builds for the `push` trigger on upstream branches had access to the secret variables for code signing but the builds for the `pull_request` trigger does not. I was wrong. The builds for both triggers have access to the secret variables for code signing. I had to push to my fork to get GH Actions to make unsigned builds.

Can you point where to download those builds? Would like to test with macos 10.14 - 10.15

Be-ing commented:
https://github.com/mixxxdj/mixxx/wiki/Testing#github-pull-requests

I've already tested with macOS 10.13.

esbrandt commented:
Only had the chance to test the signed [Mixxx-mac_sandbox_fixes-r1-b2a569c](https://github.com/mixxxdj/mixxx/suites/1723954109/artifacts/32620511) on macOS 10.14.6 yet

```bash
$ spctl --status
assessments enabled
```
It works the same as written in https://github.com/mixxxdj/mixxx/pull/3480#issuecomment-751140582 👍 
CPU load is reasonable, in contrast to DIY builds - probably due to the [older sdk ](https://github.com/mixxxdj/mixxx/pull/3420)👍 

```txt
gApp path: /Applications/Mixxx.app
app ID: org.mixxx.mixxx
version: 2.3.0
build: 2.3.0
SDK version: [none]

Usage Descriptions:

Entitlements:
com.apple.security.network.client : 1
com.apple.security.cs.allow-jit : 1
com.apple.security.device.firewire : 1
com.apple.security.app-sandbox : 1
com.apple.security.assets.music.read-write : 1
com.apple.security.device.usb : 1
com.apple.security.device.microphone : 1
com.apple.security.files.user-selected.read-write : 1
com.apple.security.files.bookmarks.app-scope : 1
com.apple.security.files.downloads.read-write : 1
com.apple.security.cs.allow-unsigned-executable-memory : 1

App signature check:
/Applications/Mixxx.app: accepted
source=Developer ID
origin=Developer ID Application: RJ Skerry-Ryan (FLYL4D545V)

codesign check:
Executable=/Applications/Mixxx.app/Contents/MacOS/mixxx
Identifier=org.mixxx.mixxx
Format=app bundle with Mach-O thin (x86_64)
CodeDirectory v=20200 size=93231 flags=0x10000(runtime) hashes=2906+5 location=embedded
Library validation warning=OS X SDK version before 10.9 does not support Library Validation
Hash type=sha256 size=32
CandidateCDHash sha256=72f179c6e97f9b845599cec9f38aab2cce28f548
Hash choices=sha256
CDHash=72f179c6e97f9b845599cec9f38aab2cce28f548
Signature size=8928
Authority=Developer ID Application: RJ Skerry-Ryan (FLYL4D545V)
Authority=Developer ID Certification Authority
Authority=Apple Root CA
Timestamp=25. Dec 2020 at 04:58:42
Info.plist entries=12
TeamIdentifier=FLYL4D545V
Sealed Resources version=2 rules=13 files=2662
Internal requirements count=1 size=176
```
Observations:
* No SDK version is provided. 
  Apple [states](https://developer.apple.com/documentation/xcode/notarizing_macos_software_before_distribution#overview) :
  > Beginning in macOS 10.15, all software built after June 1, 2019, and distributed with Developer ID must be notarized. 

  Notarization apparently works only with SDK >= 10.9. 
  I don't know how the plans are to get the final app notarized, but providing SDK version is needed.
* DMG not signed. While not necessary, signing it too can be a nice touch to help users verify the content of the disk image.
* 
Hope to have the chance to continue the test with macOS 10.15 tonight

Be-ing commented:
Thanks for confirming. What still needs confirmation is if this works on macOS 11.

> DMG not signed. While not necessary, signing it too can be a nice touch to help users verify the content of the disk image.

Are you sure? The DMG should be signed.

esbrandt commented:
> Are you sure? The DMG should be signed.

My bad

```bash
spctl -a -t open --context context:primary-signature -v ~/Mixxx-mac_sandbox_fixes-r1-b2a569c.dmg
~/Mixxx-mac_sandbox_fixes-r1-b2a569c.dmg: accepted
source=Developer ID
```

esbrandt commented:
Tested with macos10.15.7
![mixxx230-macos105-notnotarized](https://user-images.githubusercontent.com/4525897/103170690-88810700-4846-11eb-875d-7b5187361b15.png)

```bash
$ stapler validate /Applications/Mixxx.app
Processing: /Applications/Mixxx.app
Mixxx.app does not have a ticket stapled to it.
```
![mixxx230-macos105-notnotarized2](https://user-images.githubusercontent.com/4525897/103170983-cb43de80-4848-11eb-809b-7a8ae77eba9a.png)

No ticket = Not Notarized. Mixxx will not open per default on macOS 10.5+
However, you can manually add an exception in security settings.

Or by navigating to the program folder, right-clicking on **Mixxx.app**, click **Open**, then select again **Open** from the modal window. There are even more modal windows about allowing access to the **Music** folder (depends on the selected library location).

Next time Mixxx will open without warnings by double-clicking the app.

Be-ing commented:
Yes, we know it's not notorized. This PR does not try to fix that.

Holzhaus commented on src/util/sandbox.cpp:
```diff
@@ -369,80 +369,119 @@ QString Sandbox::migrateOldSettings() {
 
-    // QDir::homePath returns a path inside the sandbox.
+    // QDir::homePath returns a path inside the sandbox when running sandboxed
     QString homePath = QLatin1String("/Users/") + qgetenv("USER");
```
We cannot rely on userdata unsanitized.  Can you add a check with `qEnvironmentVariableIsEmpty`? We should also check if there are any slashes in the name.

```
$ env | grep USER=
USER=jan
$ export USER=
$ env | grep USER=
USER=
```

Be-ing replied:
I don't really understand your point. Usernames can't have slashes in them or that would probably cause all kinds of havok on any Unix system. In an extreme edge case where someone has bizarrely overriden the USER environment variable manually (I've never heard of anyone doing this and I don't think there's any reason to ever do so), this is going to fail regardless.

Holzhaus replied:
Yeah, and in that case we should display an error message and exit.

Holzhaus commented on src/util/sandbox.cpp:
```diff
@@ -369,80 +369,119 @@ QString Sandbox::migrateOldSettings() {
+                    "This only needs to be done once."
+                    "\n\n"
+                    "If you do not want to grant Mixxx access, click Cancel "
+                    "on the file picker."));
```
What happens in that case? Will mixxx start with a fresh database and config> Should we add that to the message box?

Be-ing replied:
Yeah, I'll make that more explicit.

Holzhaus commented on src/util/sandbox.cpp:
```diff
@@ -369,80 +369,119 @@ QString Sandbox::migrateOldSettings() {
-        CFRelease(url);
-        if (bookmark) {
-            QFile oldSettings(oldPath);
+        if (s_bInSandbox) {
```
This requires `Sandbox::checkSandboxed` to be called, which is not really apparent. Idea: Make `s_bIsSandboxed` a `std::optional`. If it doesn't hold a value, run `s_bIsStandboxed = Sandbox::isSandboxed()` (change `void Sandbox::checkSandbox()` to `bool Sandbox::isSandboxed` and let it return the value instead of assigning the singleton).

Be-ing replied:
Sandbox::checkSandbox is called before this.

Holzhaus replied:
Yes, but it's not obvious if someone touches the code later. I think `std::optional` would be helpful, or at least add a comment.

Be-ing replied:
I could just call Sandbox::checkSandbox at the beginning of this function, but IMO that would make the startup code a bit harder to follow.

Be-ing replied:
On further thought, the call to `Sandbox::checkSandboxed()` needs to stay where it is because `Sandbox::migrateOldSettings` is only invoked if the `--settingsPath` command line argument is not used.

Be-ing committed:
Sandbox: explain what will happen if user denies access

Be-ing committed:
Sandbox: check that home directory exists and is readable

I have no idea why this would ever not be the case. If it does happen in
some edge case, the user has much bigger problems than running Mixxx.

Be-ing commented:
Merge?

Holzhaus commented on src/util/sandbox.cpp:
```diff
@@ -369,80 +369,125 @@ QString Sandbox::migrateOldSettings() {
-    // QDir::homePath returns a path inside the sandbox.
+    // QDir::homePath returns a path inside the sandbox when running sandboxed
     QString homePath = QLatin1String("/Users/") + qgetenv("USER");
+    QDir homeDir(homePath);
```
```suggestion
    if (qEnvironmentVariableIsEmpty("USER") || qgetenv("USER").contains("/")) {
        qCritical() << "Cannot find home directory (USER environment variable invalid)";
        return QString();
    }

    QDir homeDir(homePath);
```

Be-ing committed:
Sandbox: validate $USER env variable

Co-authored-by: Jan Holthuis <jan.holthuis@ruhr-uni-bochum.de>

Be-ing commented:
Merge??

Be-ing commented:
I double checked that the latest artifact works on macOS 10.13.

Holzhaus reviewed (approved):
Windows CI test failure looks unrelated. LGTM, Thank you.
